packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js

```diff
@@ -78,14 +78,17 @@ export default createConnector({
       multiIndexContext: props.indexContextValue,
     });
 
-    this._prevState = this._prevState || {};
-
     this._cache = props.cache ? props.cache : this._cache || getInMemoryCache();
-    let cachedHits = this._cache.read({ state: searchState }) || {};
+    let cachedHits;
+    if (results === null) {
+      cachedHits = {};
+    } else {
+      cachedHits = this._cache.read({ state: results._state }) || {};
+    }
 
     if (!results) {
       return {
-        hits: extractHitsFromCachedHits(cachedHits),
+        hits: [],
```

Comment:
this isn't perfect, but I don't see how we can get something that's sync with the results when there are results, but also exists when there's no results yet (searchState is "eager", which means we shouldn't use it to cache yet, and it has a different shape to _state)